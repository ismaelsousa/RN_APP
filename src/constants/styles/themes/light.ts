/* eslint-disable @typescript-eslint/no-unused-vars */
import {DefaultTheme} from 'styled-components/native';
import * as Colors from '../colors';
import {typographyCommon, borders, spacing} from './common';

export const themeLight: DefaultTheme = {
  colors: {
    primary: {
      main: Colors.brand01,
      onMain: Colors.white,
    },
    secondary: {
      main: Colors.brand02,
      onMain: Colors.white,
    },
    error: {
      main: Colors.red,
      onMain: Colors.white,
    },
    background: {
      main: Colors.white,
      onMain: Colors.black,
    },
    surface: {
      main: Colors.black,
      onMain: Colors.white,
    },
    surface50: {
      main: Colors.grey50,
      onMain: Colors.white,
    },
    surface100: {
      main: Colors.grey100,
      onMain: Colors.white,
    },
    surface500: {
      main: Colors.grey500,
      onMain: Colors.white,
    },
    surface600: {
      main: Colors.grey600,
      onMain: Colors.white,
    },
    orange: {
      main: Colors.orange,
      onMain: Colors.white,
    },
    lightGreen: {
      main: Colors.lightGreen,
      onMain: Colors.white,
    },
  },
  typography: typographyCommon,
  borders,
  spacing,
};
