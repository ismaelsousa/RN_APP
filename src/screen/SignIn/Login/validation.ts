import * as yup from 'yup';

export const schemaLogin = yup.object({
  email: yup.string().email('Email inválido').required('Campo obrigatório'),
  password: yup.string().min(6, 'Pelos menos 6 caracteres'),
});
