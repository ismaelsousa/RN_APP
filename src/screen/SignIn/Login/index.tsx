import React from 'react';
import {Controller, useForm} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import {Alert, Platform, ScrollView, StatusBar} from 'react-native';
import {useTheme} from 'styled-components/native';
import {GoogleSignin} from '@react-native-google-signin/google-signin';

import Button from '~/components/Button';
import HeaderOptions from '~/components/HeaderOptions';
import Icon from '~/components/Icon';
import Input from '~/components/Input';
import Separator from '~/components/Separator';
import Text from '~/components/Text';
import useSignInNavigation from '~/hooks/useSignInNavigation';

import {Container, AccessText} from './styles';
import {schemaLogin} from './validation';
import BackButton from '~/components/BackButton';
import appleAuth, {
  appleAuthAndroid,
} from '@invertase/react-native-apple-authentication';
import useAuth from '~/hooks/useAuth';
import AvoidKeyboard from '~/components/AvoidKeyboard';

const Login: React.FC = () => {
  const {spacing} = useTheme();
  const navigation = useSignInNavigation();

  /**
   * Hooks
   */
  const {loading, signIn, checkIfExistUser, signUp} = useAuth();

  /**
   * Forms
   */

  const {
    control,
    handleSubmit,
    setValue,
    setFocus,
    formState: {errors},
  } = useForm({
    resolver: yupResolver(schemaLogin),
    defaultValues: {
      email: '',
      password: '',
    },
  });
  /**
   * Callback
   */
  const handleGoBack = () => navigation.goBack();

  const onSubmit = async () => {
    await handleSubmit(async ({email, password}) => {
      await signIn({email, password});
    })();
  };

  const handleAppleButton = async () => {
    // performs login request
    const appleAuthRequestResponse = await appleAuth.performRequest({
      requestedOperation: appleAuth.Operation.LOGIN,
      requestedScopes: [appleAuth.Scope.EMAIL, appleAuth.Scope.FULL_NAME],
    });

    // get current authentication state for user
    // /!\ This method must be tested on a real device. On the iOS simulator it always throws an error.
    const credentialState = await appleAuth.getCredentialStateForUser(
      appleAuthRequestResponse.user,
    );

    // use credentialState response to ensure the user is authenticated
    if (credentialState === appleAuth.State.AUTHORIZED) {
      // user is authenticated
      const hasUser = await checkIfExistUser({
        userApple: appleAuthRequestResponse.user,
      });

      if (hasUser) {
        await signIn({
          userApple: appleAuthRequestResponse.user,
        });
      } else {
        await signUp({
          ...(appleAuthRequestResponse.user
            ? {userApple: appleAuthRequestResponse.user}
            : {}),
          ...(appleAuthRequestResponse.email
            ? {email: appleAuthRequestResponse.email}
            : {}),
          ...(appleAuthRequestResponse.fullName?.familyName
            ? {lastName: appleAuthRequestResponse.fullName.familyName}
            : {}),
          ...(appleAuthRequestResponse.fullName?.givenName
            ? {firstName: appleAuthRequestResponse.fullName.givenName}
            : {}),
        });
      }
    }
  };

  const handleGoogleButton = async () => {
    try {
      const {user} = await GoogleSignin.signIn();

      const hasUser = await checkIfExistUser({
        email: user.email,
      });

      if (hasUser) {
        await signIn({
          email: user.email,
        });
      } else {
        await signUp({
          ...(user.photo ? {avatar: user.photo} : {}),
          ...(user.givenName ? {firstName: user.givenName} : {}),
          ...(user.familyName ? {lastName: user.familyName} : {}),
          email: user.email,
        });
      }
    } catch (error) {
      console.error(error);
      Alert.alert('Ops!', 'Ocorreu um erro ao realizar o login');
    }
  };

  return (
    <AvoidKeyboard>
      <Container>
        <StatusBar barStyle="dark-content" />
        <HeaderOptions
          left={<BackButton icon="closeX" onPress={handleGoBack} />}
          right={
            <Text color="primary" typography="body3">
              Esqueci minha senha
            </Text>
          }
        />
        <ScrollView showsVerticalScrollIndicator={false}>
          <Separator height={spacing.md} />
          <Text typography="h3">Login</Text>
          <Separator height={spacing.md} />
          <Controller
            control={control}
            name="email"
            render={({field: {onBlur, onChange, value, ref}}) => (
              <Input
                ref={ref}
                autoCapitalize="none"
                autoCompleteType="email"
                keyboardType="email-address"
                onChange={onChange}
                onChangeText={text => setValue('email', text)}
                value={value}
                onBlur={onBlur}
                label="Email"
                icon="checkCircle"
                iconColor="primary"
                error={errors.email?.message}
                returnKeyType="next"
                onSubmitEditing={() => setFocus('password')}
              />
            )}
          />
          <Controller
            control={control}
            name="password"
            render={({field: {onBlur, onChange, value, ref}}) => (
              <Input
                ref={ref}
                label="Senha"
                autoCapitalize="none"
                autoCompleteType="password"
                secureTextEntry
                iconColor="primary"
                onChange={onChange}
                onChangeText={text => setValue('password', text)}
                value={value}
                onBlur={onBlur}
                error={errors.password?.message}
                returnKeyType="done"
                onSubmitEditing={onSubmit}
              />
            )}
          />
          <Separator height={spacing.md} />
          <Button
            testID="button-login-email"
            loading={loading}
            disabled={loading}
            onPress={onSubmit}>
            Login
          </Button>
          <Separator height={spacing.md} />
          <AccessText color="surface500" typography="body3">
            ou acesse com login social
          </AccessText>
          <Separator height={spacing.md} />
          {(Platform.OS === 'ios' || appleAuthAndroid.isSupported) && (
            <>
              <Button
                onPress={handleAppleButton}
                typography="caption"
                icon={<Icon icon="apple" />}
                color="secondary"
                mode="outlined">
                Continuar com a Apple
              </Button>
              <Separator height={spacing.md} />
            </>
          )}
          <Button
            onPress={handleGoogleButton}
            typography="caption"
            icon={<Icon icon="google" />}
            color="secondary"
            mode="outlined">
            Continuar com o Google
          </Button>
          <Separator height={spacing.xxl} />
        </ScrollView>
      </Container>
    </AvoidKeyboard>
  );
};

export default Login;
