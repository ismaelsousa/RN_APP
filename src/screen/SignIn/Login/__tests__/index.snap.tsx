/**
 * @format
 */

import 'react-native';
import React from 'react';
import Login from '../';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {ThemeProvider} from 'styled-components/native';
import {themeDark} from '../../../../constants/styles/themes/dark';

jest.mock('@react-native-google-signin/google-signin', () => ({
  GoogleSignin: {},
}));

jest.mock('@invertase/react-native-apple-authentication', () => ({
  default: {},
  appleAuthAndroid: {},
}));

jest.mock('../../../../hooks/useSignInNavigation', () => () => ({}));

it('renders correctly', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={themeDark}>
        <Login />
      </ThemeProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
