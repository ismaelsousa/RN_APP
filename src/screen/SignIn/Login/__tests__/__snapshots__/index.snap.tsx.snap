// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "flexGrow": 1,
        },
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginBottom": 0,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 0,
        },
      ]
    }
  >
    <View
      height={20}
      style={
        Array [
          Object {
            "height": 20,
            "maxHeight": 20,
            "minHeight": 20,
          },
        ]
      }
      testID="Separator"
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "maxWidth": "100%",
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Image
          resizeMode="contain"
          size={15}
          source={
            Object {
              "testUri": "../../../src/assets/icons/closeX/Vector.png",
            }
          }
          style={
            Array [
              Object {
                "maxHeight": 15,
                "maxWidth": 15,
                "minHeight": 15,
                "minWidth": 15,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
        testID="Separator"
      />
      <Text
        color="primary"
        style={
          Array [
            Object {
              "color": "#3C9BA6",
              "fontFamily": "OpenSans-Regular",
              "fontSize": 14,
            },
          ]
        }
        typography="body3"
      >
        Esqueci minha senha
      </Text>
    </View>
    <View
      height={20}
      style={
        Array [
          Object {
            "height": 20,
            "maxHeight": 20,
            "minHeight": 20,
          },
        ]
      }
      testID="Separator"
    />
    <RCTScrollView>
      <View>
        <Text
          color="surface"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "Montserrat-Bold",
                "fontSize": 32,
              },
            ]
          }
          typography="h3"
        >
          Login
        </Text>
        <View
          height={20}
          style={
            Array [
              Object {
                "height": 20,
                "maxHeight": 20,
                "minHeight": 20,
              },
            ]
          }
          testID="Separator"
        />
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            color="surface600"
            style={
              Array [
                Object {
                  "color": "#41474D",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 14,
                },
                Object {
                  "marginBottom": 10,
                },
              ]
            }
            typography="body3"
          >
            Email
          </Text>
          <View
            borderColor="#6B7785"
            color="transparent"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#6B7785",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingBottom": 0,
                  "paddingLeft": 15,
                  "paddingRight": 15,
                  "paddingTop": 0,
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCompleteType="email"
              keyboardType="email-address"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              rejectResponderTermination={true}
              returnKeyType="next"
              style={
                Array [
                  Object {
                    "color": "#000",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontFamily": "OpenSans-Regular",
                    "fontSize": 14,
                    "minHeight": 48,
                  },
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
            <View
              iconPosition="right"
              style={
                Array [
                  Object {
                    "paddingLeft": 10,
                    "paddingRight": 0,
                  },
                ]
              }
            >
              <Image
                resizeMode="contain"
                size={20}
                source={
                  Object {
                    "testUri": "../../../src/assets/icons/checkCircle/Vector.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "maxHeight": 20,
                      "maxWidth": 20,
                      "minHeight": 20,
                      "minWidth": 20,
                    },
                    Object {
                      "tintColor": "#3C9BA6",
                    },
                    undefined,
                  ]
                }
              />
            </View>
          </View>
        </View>
        <View
          height={10}
          style={
            Array [
              Object {
                "height": 10,
                "maxHeight": 10,
                "minHeight": 10,
              },
            ]
          }
          testID="Separator"
        />
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            color="surface600"
            style={
              Array [
                Object {
                  "color": "#41474D",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 14,
                },
                Object {
                  "marginBottom": 10,
                },
              ]
            }
            typography="body3"
          >
            Senha
          </Text>
          <View
            borderColor="#6B7785"
            color="transparent"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#6B7785",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "paddingBottom": 0,
                  "paddingLeft": 15,
                  "paddingRight": 15,
                  "paddingTop": 0,
                },
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCompleteType="password"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              rejectResponderTermination={true}
              returnKeyType="done"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontFamily": "OpenSans-Regular",
                    "fontSize": 14,
                    "minHeight": 48,
                  },
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                iconPosition="right"
                style={
                  Array [
                    Object {
                      "paddingLeft": 10,
                      "paddingRight": 0,
                    },
                  ]
                }
              >
                <Image
                  resizeMode="contain"
                  size={20}
                  source={
                    Object {
                      "testUri": "../../../src/assets/icons/openEye/Vector.png",
                    }
                  }
                  style={
                    Array [
                      Object {
                        "maxHeight": 20,
                        "maxWidth": 20,
                        "minHeight": 20,
                        "minWidth": 20,
                      },
                      Object {
                        "tintColor": "#3C9BA6",
                      },
                      undefined,
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          height={20}
          style={
            Array [
              Object {
                "height": 20,
                "maxHeight": 20,
                "minHeight": 20,
              },
            ]
          }
          testID="Separator"
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#3C9BA6",
              "borderColor": "#3C9BA6",
              "borderRadius": 10,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 12,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 12,
            }
          }
          testID="button-login"
        >
          <Text
            color="#fff"
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#fff",
                  "fontFamily": "OpenSans-Bold",
                  "fontSize": 16,
                },
              ]
            }
            typography="body2"
          >
            Login
          </Text>
        </View>
        <View
          height={20}
          style={
            Array [
              Object {
                "height": 20,
                "maxHeight": 20,
                "minHeight": 20,
              },
            ]
          }
          testID="Separator"
        />
        <Text
          color="surface500"
          style={
            Array [
              Object {
                "color": "#6B7785",
                "fontFamily": "OpenSans-Regular",
                "fontSize": 14,
              },
              Object {
                "alignSelf": "center",
              },
            ]
          }
          typography="body3"
        >
          ou acesse com login social
        </Text>
        <View
          height={20}
          style={
            Array [
              Object {
                "height": 20,
                "maxHeight": 20,
                "minHeight": 20,
              },
            ]
          }
          testID="Separator"
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#072B59",
              "borderRadius": 10,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 12,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 12,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "left": 20,
                  "position": "absolute",
                },
              ]
            }
          >
            <Image
              resizeMode="contain"
              size={20}
              source={
                Object {
                  "testUri": "../../../src/assets/icons/google/Vector.png",
                }
              }
              style={
                Array [
                  Object {
                    "maxHeight": 20,
                    "maxWidth": 20,
                    "minHeight": 20,
                    "minWidth": 20,
                  },
                ]
              }
            />
          </View>
          <Text
            color="#072B59"
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "#072B59",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 16,
                },
              ]
            }
            typography="caption"
          >
            Continuar com o Google
          </Text>
        </View>
        <View
          height={45}
          style={
            Array [
              Object {
                "height": 45,
                "maxHeight": 45,
                "minHeight": 45,
              },
            ]
          }
          testID="Separator"
        />
      </View>
    </RCTScrollView>
  </RNCSafeAreaView>
</View>
`;
